extensions/amp-a4a/0.1/signature-verifier.js

```diff
@@ -0,0 +1,207 @@
+import {xhrFor} from '../../../src/xhr';
+
+/** @enum {number} */
+export const VerificationFailure = {
+  NO_FAULT: 1,
+  KEY_NOT_FOUND: 2,
+  SIGNATURE_MISMATCH: 3,
+};
+
+/**
+ * @param {!Window} win
+ * @return {!SignatureVerifier}
+ */
+export function signatureVerifierFor(win) {
+  const propertyName = 'AMP_FAST_FETCH_SIGNATURE_VERIFIER_';
+  if (!win[propertyName]) {
+    win[propertyName] = new SignatureVerifier(win);
+  }
+  return win[propertyName];
+}
+
+class SignatureVerifier {
+  /**
+   * @param {!Window} win
+   * @private
+   */
+  constructor(win) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /**
+     *
+     * @private {?Object<string, {promise: !Promise<boolean>, keys: !Object<string, ?Promise<?CryptoKey>>}>}
+     */
+    this.signers_ = cryptoFor(win).isPkcsAvailable() ? {} : null;
+  }
+
+  /**
+   * @param {string} signingServiceName
+   */
+  loadKeyset(signingServiceName) {
+    if (this.signers_ && !this.signers_[signingServiceName]) {
+      const keys = {};
+      const promise = this.fetchAndAddKeys_(keys, signingServiceName, null);
+      this.signers_[signingServiceName] = {promise, keys};
+    }
+  }
+
+  /**
+   * @param {string} signingServiceName
+   * @param {string} keypairId
+   * @param {!ArrayBuffer} signature
+   * @param {!ArrayBuffer} creative
+   * @return {!Promise<?VerificationFailure>)
+   */
+  verify(signingServiceName, keypairId, signature, creative) {
+    if (this.signers_) {
+      const signer = this.signers_[signingServiceName];
+      return signer.promise.then((success) => {
+        if (success) {
+          const keyPromise = signer.keys[keypairId];
+          if (keyPromise === undefined) {
+            signer.promise = this.fetchAndAddKeys_(
+                                     signer.keys, signingServiceName, keypairId)
+                                 .then((success) => {
+                                   if (!signer.keys[keypairId]) {
+                                     signer.keys[keypairId] = null;
+                                   }
+                                   return success;
+                                 });
+            return this.verify(
+                signingServiceName, keypairId, signature, creative);
+          } else if (keyPromise === null) {
+            return VerificationFailure.KEY_NOT_FOUND;
+          } else {
+            return keyPromise.then((key) => {
+              if (key) {
+                return cryptoFor(this.win_)
+                    .verifyPkcs(key, signature, creative)
+                    .then(
+                        (result) => {
+                          if (result) {
+                            return null;
+                          } else {
+                            return VerificationFailure.SIGNATURE_MISMATCH;
+                          }
+                        },
+                        (err) => {
+                          dev().error(`Failed to verify signature: ${
```

Comment:
This is clang-format's fault. Apparently other people within Google have also complained about this behavior. I am not totally sure what the right thing to do about it is, since I don't want to have something that the autoformatter will just revert later.